src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/HeatmapOfflineAggregator.scala

```diff
@@ -0,0 +1,99 @@
+import com.microsoft.partnercatalyst.fortis.spark.sinks.cassandra.dto.{ComputedTile, Event, HeatmapTile, PopularPlace}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, SaveMode, SparkSession}
+
+class HeatmapOfflineAggregator(session: SparkSession) extends OfflineAggregator[HeatmapTile] {
+  private val ComputedTilesTable = "computedtiles"
+
+  override def aggregate(events: RDD[Event]): RDD[HeatmapTile] = {
+    val tiles = events.flatMap(event=>{
+      Seq(
+        event,
+        event.copy(externalsourceid = "all"),
+        event.copy(pipelinekey = "all", externalsourceid = "all")
+      )
+    }).flatMap(CassandraHeatmapTiles(_))
+    tiles.keyBy(r=>{(
+      r.heatmaptileid, r.period,
+      r.periodtype, r.perioddate,
+      r.conjunctiontopic1, r.conjunctiontopic2, r.conjunctiontopic3,
+      r.tileid, r.tilez,
+      r.pipelinekey, r.externalsourceid
+    )}).reduceByKey((a,b)=>{
+      val mentionCount = a.mentioncount + b.mentioncount
+      val sentimentNumerator = a.avgsentimentnumerator + b.avgsentimentnumerator
+      a.copy(
+        mentioncount = mentionCount,
+        avgsentimentnumerator = sentimentNumerator
+      )
+    }).values
+  }
+
+  private def aggregateTileBuckets(tiles: RDD[HeatmapTile], keyspace: String): DataFrame = {
+    import session.implicits._
+    val tilesComputed = tiles.keyBy(r=>{(
+      r.period,
+      r.periodtype, r.perioddate,
+      r.conjunctiontopic1, r.conjunctiontopic2, r.conjunctiontopic3,
+      r.tileid, r.tilez,
+      r.pipelinekey, r.externalsourceid
+    )}).reduceByKey((a,b)=>{
+      val mentionCount = a.mentioncount + b.mentioncount
+      val sentimentNumerator = a.avgsentimentnumerator + b.avgsentimentnumerator
+      a.copy(
+        mentioncount = mentionCount,
+        avgsentimentnumerator = sentimentNumerator
+      )
+    }).values.map(CassandraTileBucket(_))
+
+    val tileBucketDF = tilesComputed.toDF()
+
+    tileBucketDF.createOrReplaceTempView("aggregatedtiles")
+
+    val computedTilesSourceDF = session.read.format("org.apache.spark.sql.cassandra")
+      .options(Map("keyspace" -> keyspace, "table" -> ComputedTilesTable))
+      .load()
+
+    computedTilesSourceDF.createOrReplaceTempView(ComputedTilesTable)
+
+    val cassandraSave = session.sql(IncrementalUpdateQuery)
+
+    cassandraSave
+  }
+
+  private def IncrementalUpdateQuery: String = {
+    val GroupedBaseColumnNames = Seq("periodtype", "period", "conjunctiontopic1", "conjunctiontopic2", "conjunctiontopic3", "perioddate", "tileid", "tilez", "pipelinekey", "externalsourceid").mkString(",a.")
+
+    s"SELECT a.$GroupedBaseColumnNames, " +
+    s"       a.mentioncount + IF(IsNull(b.mentioncount), 0, b.mentioncount) as mentioncount, " +
+    s"       a.avgsentimentnumerator + IF(IsNull(b.avgsentimentnumerator), 0, b.avgsentimentnumerator) as avgsentimentnumerator " +
+    s"FROM   aggregatedtiles a " +
+    s"LEFT OUTER JOIN ${ComputedTilesTable} b " +
```

Comment:
Would it be possible to turn the bigint and avgsentimentnumerator in the computedtiles table into counter columns so that we don't have to do this read here? Read-before-write was causing us slowness in the PR that JC just submitted (https://github.com/CatalystCode/project-fortis-spark/pull/120) so we may want to avoid it here too.

Granted, tables with counter columns can't be used in materialized views, but we can also just replace those with normal tables and manage the data replication ourselves by just adding a few more writes from Spark to Cassandra.

Just something to think about.

Reply:
Unfortunately that is not possible. Materialized views can contain any counter columns in Cassandra. We have 3 MVs that are dependent on the `mentioncount ` field.

Reply:
Would it be possible to replace the materialized views with straight tables?

Reply:
not possible as we need the `computedtiles` dataset sorted by `mentioncount` to support our top 5 sources / locations / terms GQL queries.

Reply:
Cassandra doesnt allow you to add a counter column to your PK